rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+| action       | initial state | change                           |
+| ------------ | ------------- | -------------------------------- |
+| role removal | _admin_       | _admin_ tries to remove own role |
+
+The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by manually instrumenting the auth process using tctl.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
+2. User with a role that grants user/role editing capabilities is _admin_.
```

Comment:
Are we only considering local users here? It seems that way from the implementation PR, should make that clear if so.

It seems that with this design we will be explicitly forcing the existence of a local user with role edit permissions? I could see users wanting to avoid this so that everything MUST go through SSO.

Many existing clusters will not have this property when upgrading, and newly created clusters have zero users so they will also not satisfy this property. We should be careful about what to do in this scenario so that SSO admin users and local `tctl` are not blocked from editing roles (or any other actions)

Reply:
> Are we only considering local users here?

Taking only local users into consideration will ensure that in case of SSO users deletion / SSO disconnection there will be still someone able to restore the cluster.

> Many existing clusters will not have this property when upgrading, and newly created clusters have zero users so they will also not satisfy this property. We should be careful about what to do in this scenario so that SSO admin users and local tctl are not blocked from editing roles (or any other actions)

Good point! We need to consider this. I think is safe to assume that currently there are clusters with only SSO users and they don't want local users. 
I think there can be two approaches here:
1. treat the SSO users as regular users. This will allow lock situations when destructive changes are made on the SSO provider side (disconnect, user deletion, group deletion, etc...). We can't control them. We need to assume that 
if a connector is registered and grants access to a role that allows managing the cluster this means we are also in a valid state.